### PR TITLE
HPCC-24181 Thor startup time stat.

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review.

Separate issue, how are these best exposed in Eclwatch?
Should all global workunit properties be viewable in a property view. I'm not sure how the ones currently visible are picked..

**Comment by richardkchapman** [comment]
Unless I'm reading it wrong, this measures the time from thor process starting to "being ready" but ignores any time take, for example, waiting for kubernetes to start the thormaster pod ?

File: system/jlib/jstatcodes.h [hunk 1]
```diff
@@ -233,6 +233,8 @@ enum StatisticKind
     StTimeBlocked,
     StCycleBlockedCycles,
     StCostExecute,
+    StTimeThorStartup,                  // Time taken for Thor instance to spin up before beginning to run query
```
Comment (ghalliday) [inline system/jlib/jstatcodes.h:236]: I'm not sure these should have the name startup in them.  Should it be TimeStartup (which can then be reused in other contexts)?

File: thorlcr/master/thmastermain.cpp [hunk 2]
```diff
@@ -954,6 +956,18 @@ int main( int argc, const char *argv[]  )
...
             // NB: workunit/graphName only set in one-shot mode (if isCloud())
+
+            if (workunit)
+            {
+                Owned<IWorkUnitFactory> factory = getWorkUnitFactory();
+                Owned<IConstWorkUnit> wuRead = factory->openWorkUnit(workunit);
+                if (wuRead)
+                {
+                    Owned<IWorkUnit> wuWrite = &wuRead->lock();
+                    updateWorkunitStat(wuWrite, SSTglobal, nullptr, StTimeThorStartup, nullptr, thorStartupTimer.elapsedNs());
```
Comment (ghalliday) [inline thorlcr/master/thmastermain.cpp:967]: I think this should be updating the startup time for the graph, rather than global.

**Comment by ghalliday** [comment]
To include the k8s startup time it would need to record that information in the queue entry, or in the workunit.  Alternatively it could add a stat StWhenQueue (graph), and then base it on that.  (I'm beginning to wonder if we need a more compact global stats storage format.)

That would require clocks to be (reasonably) well synchronized between machines..

**Comment by richardkchapman** [comment]
The time between being added to a queue and removed from the queue might be something of general interest - and could potentially be tracked by dali using just its local clock?

**Comment by ghalliday** [comment]
Yes, and might link in with the metrics that ken wants to collect.

**Comment by ghalliday** [comment]
@jakesmith forgot to assign back to you

**Comment by jakesmith** [comment]
Closing for now. It needs rethinking to capture the total time and probably move timing out of global and make it per graph.

**Pull request closed without merging**